tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,300 @@
+    pub async fn get_or_init<F>(&self, f: F) -> &T
+    where
+        F: Fn() -> Pin<Box<dyn Future<Output = T> + Send>>,
+    {
+        if self.initialized() {
+            // SAFETY: once the value is initialized, no mutable references are given out, so
+            // we can give out arbitrarily many immutable references
+            return unsafe { self.get_unchecked() };
+        } else {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or current thread is awoken after another thread
+            // has intialized the value and closed the semaphore, in which case self.initialized
+            // is true and we don't set the value here
+            match self.sema.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        let value = f().await;
+
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        // It's important that we close the semaphore only after set_to_initialized call
+                        // otherwise a waiting thread would be woken up and find an uninitialized
+                        // state after failing to acquire a permit, which causes a panic
+                        self.set_to_initialized();
+                        self.sema.close();
+
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        panic!("acquired semaphore after value was already initialized");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        panic!("semaphore acquire call failed when value not already intialized");
+                    }
+                }
+            }
+        }
+    }
+}
+
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::get`] and [`OnceCell::set`] methods
+///
+/// [`OnceCell::get`]: crate::sync::OnceCell::get
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub enum OnceCellError {
+    /// The value in OnceCell was already initialized
+    AlreadyInitialized,
+
+    /// The value in OnceCell was not previously initialized
+    NotInitialized,
+}
```

Comment:
I would like to split this into two error types, as the caller otherwise ends up with extra unreachable branches if they match on it.